### PR TITLE
Handle undefined content download in video enrichment

### PR DESCRIPTION
**Pull request by andygout**
### Description
On Wed 30 Nov 2022 at 05:52, this app crashed, reporting the following healthcheck failure in the [#ft-next-support](https://financialtimes.slack.com/archives/C042NBBTM) Slack channel (see Slack [post](https://financialtimes.slack.com/archives/C042NBBTM/p1669788148788669)):

> :warning: Heroku H10: ft-next-syndication-api is stable, severity 1 issue

The Splunk logs showed that the cause was this unhandled error triggered in `/server/lib/enrich/video.js` owing to `content.download` being `undefined`:

```
{ [-]
   app: { [-]
     commit: cb20ac98682c6b676dc9da69868bb714644e5439
     name: next-syndication-api
     nodeVersion: 16.18.1
     region: EU
     releaseDate: 2022-11-24T12:29:56Z
   }
   error: { [-]
     cause: null
     code: UNKNOWN
     data: { [+]
     }
     isOperational: false
     message: Cannot set properties of undefined (setting 'extension')
     name: TypeError
     relatesToSystems: [ [-]
     ]
     stack: TypeError: Cannot set properties of undefined (setting 'extension')
    at Function.video (/app/server/lib/enrich/video.js:18:29)
    at enrich (/app/server/lib/enrich/index.js:14:34)
    at /app/server/lib/get-content.js:19:31
    at Array.filter (<anonymous>)
    at exports (/app/server/lib/get-content.js:19:16)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async exports (/app/server/controllers/resolve.js:42:14)
     statusCode: null
   }
   event: UNHANDLED_ERROR
   level: error
   message: TypeError: Cannot set properties of undefined (setting 'extension')
}
```

I identified that a content item of type `video` with an empty `attachments` array would cause this error and so I needed to find a sample document that could cause this error and add it as a test fixture.

This Elasticsearch query (run against our main content index) shows the different number of `attachments` array items that content items of type `video` have, with 10 content IDs for each:

```
{
    "query": {
        "term": {
            "type": "video"
        }
    },
    "aggs": {
        "annotations_length": {
            "terms": {
                "script": "int instance = 0; List copy = new ArrayList(params._source.attachments); copy.removeIf(i -> i.mediaType != 'video/mp4'); for (int i = 0; i < copy.length; ++i) { instance++; } return instance;"
            },
            "aggs": {
                "ids": {
                    "terms": {
                        "field": "id",
                        "size": 10
                    }
                }
            }
        }
    },
    "size": 0
}

```

Results:

```
{
    …
    "hits": {
        "total": {
            "value": 10000,
            "relation": "gte"
        },
        "max_score": null,
        "hits": []
    },
    "aggregations": {
        "annotations_length": {
            "doc_count_error_upper_bound": 0,
            "sum_other_doc_count": 0,
            "buckets": [
                {
                    "key": "1",
                    "doc_count": 11971,
                    "ids": {
                        "doc_count_error_upper_bound": 4,
                        "sum_other_doc_count": 11961,
                        "buckets": [
                            {
                                "key": "00012aaa-bbcd-3643-a485-81584f40352b",
                                "doc_count": 1
                            },
                            …[9 more]
                        ]
                    }
                },
                {
                    "key": "0",
                    "doc_count": 5355,
                    "ids": {
                        "doc_count_error_upper_bound": 4,
                        "sum_other_doc_count": 5345,
                        "buckets": [
                            {
                                "key": "0002839e-6dd8-34ff-ae14-e5dfb40b2eff",
                                "doc_count": 1
                            },
                            …[9 more]
                        ]
                    }
                },
                {
                    "key": "3",
                    "doc_count": 4391,
                    "ids": {
                        "doc_count_error_upper_bound": 4,
                        "sum_other_doc_count": 4381,
                        "buckets": [
                            {
                                "key": "001424fa-57fd-4cef-90b6-dcf030e916d9",
                                "doc_count": 1
                            },
                            …[9 more]
                        ]
                    }
                }
            ]
        }
    }
}
```

However, I noticed that unit tests are not currently running as part of the CI process, though the effect of adding the new content fixture is as follows:
- Commit: [Enable unit tests in CI `ae0d7c8`](https://github.com/Financial-Times/next-syndication-api/commit/ae0d7c8d6b861a8be1f0947fd1c9991f3766790a)
  - [CI process :tool-kit/test step](https://app.circleci.com/pipelines/github/Financial-Times/next-syndication-api/2372/workflows/acaad48c-2db8-42d8-99ba-bf64dd402182/jobs/9822) - fails owing to existing broken unit test
- Commit: [Add test case for video content item with empty attachments array `b63835d`](https://github.com/Financial-Times/next-syndication-api/commit/b63835df8fd253db7820d0b39553e9983f9c6b84)
  - [CI process :tool-kit/test step](https://app.circleci.com/pipelines/github/Financial-Times/next-syndication-api/2373/workflows/fb028b21-3fe2-4163-98f5-1e948e3dd43d/jobs/9830) - fails owing to the new test case, with the same error as was observed in the Splunk logs: `TypeError: Cannot set properties of undefined (setting 'extension')`
- Commit: [Add handling for undefined content.download in video enrichment `c1f70fa`](https://github.com/Financial-Times/next-syndication-api/commit/c1f70fa82dccdf7328f5b4fa971d94418ad66878)
  - [CI process :tool-kit/test step](https://app.circleci.com/pipelines/github/Financial-Times/next-syndication-api/2375/workflows/cb762a94-c9de-400e-8f48-e26a5393a966/jobs/9840) - fails owing to existing broken unit test, i.e. same as first commit, indicating that the new test case is passing
- Commit: [Disable unit tests in CI](https://github.com/Financial-Times/next-syndication-api/commit/1849dd6d68c458644e69ded80783767178fb2350)
  - [CI process :tool-kit/test step](https://app.circleci.com/pipelines/github/Financial-Times/next-syndication-api/2377/workflows/d45d4700-906b-4a85-b6b1-61f708c25f7f/jobs/9852) - passes because the unit tests are no longer run as part of the CI process (which should be addressed in a separate commit)

### Ticket
N/A

### What is the new version number in package.js?

Can I leave this with you please?

### Link to the next-syndication-dl PR which bumps the next-syndication-api version

Can I leave this with you please?

### CODE REVIEW

File: server/lib/enrich/video.js [hunk 1]
```diff
@@ -15,7 +15,10 @@ module.exports = exports = function video(content, contract, graphicSyndicationF
 	content.extension = DOWNLOAD_ARCHIVE_EXTENSION;
 
 	content.download = Array.from(content.attachments).reverse().find(item => item.mediaType === DOWNLOAD_MEDIA_TYPES.video);
-	content.download.extension = (content.download) ? mime.extension(content.download.mediaType) : '';
+
+	if (content.download) {
+		content.download.extension = (content.download) ? mime.extension(content.download.mediaType) : '';
```
Comment (rowanmanning) [inline server/lib/enrich/video.js:20]: As this has moved into a condition we can guarantee that `content.download` is truthy, so the ternary is redundant now:

```suggestion
		content.download.extension = mime.extension(content.download.mediaType);
```

File: server/lib/enrich/video.js [hunk 2]
```diff
@@ -15,7 +15,10 @@ module.exports = exports = function video(content, contract, graphicSyndicationF
...
+	}
```
Comment (AniaMakes) [inline server/lib/enrich/video.js:21]: [Observation] I need to admit, I'm not sure what happens with `content.download.extension`. Because currently you only set the extension if the `content.download` is true, and we'd end up with `content.download.extension` of `undefined` which may or may not be a problem - I don't know off hand. 

Looking at this code, I'm a little bit confused as to how it works. So `content.download` is an array, right? But in the next line (in the old version) we're treating `content.download` as an object - so which one do we want it to be - an array, or an object?

Comment (andygout) [reply server/lib/enrich/video.js:21]: `content-attachments` is an array that can be empty (which currently causes an error) or, if populated, will look something like this:

```js
[
  {
    "url": "https://next-media-api.ft.com/renditions/15053038942430/640x360.mp4",
    "width": 640,
    "height": 360,
    "duration": 178943,
    "codec": "h264",
    "mediaType": "video/mp4"
  },
  {
    "url": "https://next-media-api.ft.com/renditions/15053038942430/1280x720.mp4",
    "width": 1280,
    "height": 720,
    "duration": 178943,
    "codec": "h264",
    "mediaType": "video/mp4"
  },
  {
    "url": "https://next-media-api.ft.com/renditions/15053038942430/1920x1080.mp4",
    "width": 1920,
    "height": 1080,
    "duration": 178943,
    "codec": "h264",
    "mediaType": "video/mp4"
  },
  {
    "mediaType": "text/vtt",
    "url": "https://next-media-api.ft.com/captions/15053038942430.vtt"
  }
]
```

`content.download`, providing a suitable object in the array can be found, will be an object (both before and after these changes).

Ref. [MDN Web Docs: Array.prototype.find()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find)
  > The `find()` method returns the first element in the provided array that satisfies the provided testing function. If no values satisfy the testing function, [undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined) is returned.

Comment (AniaMakes) [reply server/lib/enrich/video.js:21]: Agh, yes, I was looking at the `Array.from`. Chaining has its place, but can also be confusing! 😅

**Review by rowanmanning — APPROVED** [review]

**Review by AniaMakes — APPROVED** [review]
I was talking with Tatiana earlier how I'd make the `content.download` default to an empty string, but I think this will do the job too, and may in fact be better overall now that I think about it.

**Comment by andygout** [comment]
I forgot to do these bits - looking into it now:

### What is the new version number in package.js?
> If you have made any changes other than documentation, you need to follow a special deploy process, as described here https://github.com/Financial-Times/next-syndication-dl#deploying-the-download-server

### Link to the next-syndication-dl PR which bumps the next-syndication-api version
> Add link to the PR
